BUGFIX: Uncached ConfigurationManager

mhsdesign opened the pull request:
### BUGFIX: Proper uncached configurationManager mode

It is purposely not allowed to disable the cache at runtime (when you have a configuration manager at hand)

The usage to create a configuration manager without caching, you need to have your own request handler and boot only this step:

```php
Scripts::initializeConfiguration($this->bootstrap, false);
```

---

### BUGFIX: ConfigurationManager with disabled cache doesn't replace environment variables in setting

@kitsunet and me need this for building https://github.com/neos/setup/pull/59 at super early boot time - pre compile time.

We want to use the config manager without cache, which currently has a bug and you cant really disable the cache unless using reflection to set `temporaryDirectoryPath` to null.

The config manager with disabled cache doesnt replace environment variables in settings.


<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked wit `!!!` and have upgrade-instructions

mhsdesign committed:
BUGFIX: ConfigurationManager with disabled cache doesn't replace environment variables in settings

(no cache mode is technically possible by not setting $temporaryDirectoryPath via `setTemporaryDirectoryPath`)

kitsunet commented on Neos.Flow/Classes/Configuration/ConfigurationManager.php:
```diff
@@ -423,6 +423,11 @@ protected function processConfigurationType(string $configurationType)
             $this->replaceConfigurationForConfigurationType($configurationType, $cachePathAndFilename);
             @unlink($cachePathAndFilename);
+        } else {
+            // in case the cache is disabled (implicitly, because temporaryDirectoryPath is null)
```
This was in fact never a planned use case ;)

mhsdesign replied:
```
    /**
     * An absolute file path to store configuration caches in. If null no cache will be active.
     *
     * @var string
     */
    protected $temporaryDirectoryPath;
```

this has been there since EVER!

mhsdesign replied:
you build it ^^ more or less accidental: https://github.com/neos/flow-development-collection/commit/85a7f15664cca6dfba7911b1ed8ee34cc2296102

mhsdesign replied:
and later the feature Runtime evaluation of env and constants in Configuration was added https://github.com/neos/flow-development-collection/commit/1ca7951743ac780bfb6edb83435532eafabd2ef3#diff-83398c2914c2ce46b6e55c9821468814652a275d332692492bb6edbbb84607f9R617

kitsunet replied:
Yep, I know, and I _think_ there was some low level use case at that time that required it to work that way. I would rather close that road than open it :D

kitsunet replied:
Although, the more I think about it, why even not, might be useful to have this in the long run, I really don't like it but well...

kitsunet commented on Neos.Flow/Classes/Configuration/ConfigurationManager.php:
```diff
@@ -423,6 +423,11 @@ protected function processConfigurationType(string $configurationType)
+            // in case the cache is disabled (implicitly, because temporaryDirectoryPath is null)
+            // we need to still handle replacing the environment variables like `%FLOW_PATH_ROOT%` in the configuration
+            $configuration = $this->unprocessedConfiguration[$configurationType];
+            $this->configurations[$configurationType] = eval('return ' . $this->replaceVariablesInPhpString(var_export($configuration, true)) . ';');
```
mmmmmmmmmmmm, ev(i|a)l. Do we really need this for the checks? I guess so 🤔

mhsdesign committed:
BUGFIX: Proper uncached configurationManager mode

It is purposely not allowed to disable the cache at runtime (when you have a configuration manager at hand)

The usage to create a configuration manager without caching, you need to have your own request handler and boot only this step:

```
Scripts::initializeConfiguration($this->bootstrap, false);
```

mhsdesign committed:
WE HATE PSALM

mhsdesign committed:
TASK: Use PolicyLoader without cache if ConfigurationManager is also without cache

kitsunet commented:
Not beautiful, but lets us use the health check really low level and fast. Might be useful in other situations as well...

mhsdesign commented:
Last psalm failure is unrelated

mhsdesign committed:
TASK: Guard against `temporaryDirectoryPath == null`

mhsdesign committed:
TASK: Try to fix test

mhsdesign committed:
TASK: Test configuration manager with disabled cache

fcool commented on Neos.Flow/Classes/Configuration/ConfigurationManager.php:
```diff
@@ -152,11 +152,11 @@ class ConfigurationManager
+     * An absolute file path to store configuration caches in. If not set, no cache will be active.
      *
-     * @var string
+     * @var ?string
```
this is no valid php doctype - in doctypes there are only union types - so there it should be `string|null`

mhsdesign replied:
well i didn't know that actually thanks ^^

mhsdesign replied:
```suggestion
     * @var string|null
```

fcool reviewed (approved):
Found a minor issue in a phpdoc annotation - I would be fine with ignoring it - but even better would be probably to have it "correct"

mhsdesign committed:
Fix php doc
